### DISCUSSION TITLE
Questions about user identification

### DISCUSSION THREAD

**Discussion by dyang** [TelemetryDeck/Viewer#240] [General]
Hi,

I understand TelemetryDeck's focus on privacy, and that it uses device identification as a base to identify users by default. I have three questions:

1. In case users use the same app on multiple devices, e.g. iPhone and iPad, does each device count as a unique user?
2. If a user deletes an app and later reinstalls it on the same device, will they be recognized as the same user or a new user?
3. If a user switches to a different device, e.g. by upgrading to a new phone, will they be recognized as a new user?

Thanks!

**Comment by winsmith** [comment]
Hi, you'll need to distinguish between default behavior and behavior with a specified user ID here.

By default, the TelemetryDeck SDKs will generate a user ID for you based on the `identifierForVendor` where available, or a random UUID otherwise. In this case, all three questions can be answered with "this counts as a new user", since the user ID will be different for each new installation of the app.

However, you can set your own user identifier, like the email address or internal user ID or login name of the user. This will be salted and hashed on device before passing it on to the server. Since your user uses the same email address regardless of the device, the generated ID will be the same as well. 

If your app doesn't have login or something similar, you could also generate a random UIUD, sync that via iCloud or CloudKit, and then use the same ID in all instances of the app. This will also result in the user being recognized between various instances of the app.

**Reply by dyang** [reply]
@winsmith Thank you Daniel for the explanation. 

In case of generating an app-specific id and store it in iCloud/CloudKit, is there an API to set the custom user id? e.g. `TelemetryManager.shared.setUserIdentifier(xxx)` or something similar? In one of my CloudKit apps, syncing data to a new device might take up to a minute, during which the app doesn't know what the custom id is.

**Reply by winsmith** [reply]
There is! You can use [`TelemetryClient.configuration.defaultUser`](https://github.com/TelemetryDeck/SwiftClient/blob/main/Sources/TelemetryClient/TelemetryClient.swift#L44) to set a user identifier, and the change will take effect for all signals generated from then on. 

I recommend also disabling analytics while you're still waiting for iCloud to sync using [`TelemetryClient.confiuration.analyticsDisabled`](https://github.com/TelemetryDeck/SwiftClient/blob/main/Sources/TelemetryClient/TelemetryClient.swift#L113) because otherwise you'll get duplicate entries for users before and after the sync.

**Reply by dyang** [reply]
Thanks Daniel, this is great to know.
